snapshot-cassandra/src/main/scala/com/evolutiongaming/kafka/journal/snapshot/cassandra/SnapshotCassandra.scala

```diff
@@ -0,0 +1,172 @@
+    numberOfSnapshots: Int
+  ): F[SnapshotStoreFlat[F]] =
+    for {
+      schema <- SetupSnapshotSchema[F](schemaConfig, origin, consistencyConfig)
+      statements <- Statements.of[F](schema, consistencyConfig)
+    } yield SnapshotCassandra(statements, numberOfSnapshots)
+
+  private sealed abstract class Main
+
+  def apply[F[_]: MonadThrow](statements: Statements[F], numberOfSnapshots: Int): SnapshotStoreFlat[F] = {
+    new Main with SnapshotStoreFlat[F] {
+
+      def save(key: Key, snapshot: SnapshotRecord[EventualPayloadAndType]): F[Unit] = {
+        statements.selectMetadata(key).flatMap {
+          // such snapshot is already saved, overwrite
+          case s if s.values.exists { case (seqNr, _) => snapshot.snapshot.seqNr == seqNr } =>
+            update(key, s, snapshot)
+          // there is a free place to add a snapshot
+          case s if s.size < numberOfSnapshots => insert(key, s, snapshot)
+          // all rows are taken, we have to update one of them
+          case s => replace(key, s, snapshot)
+        }
+      }
+
+      def insert(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        snapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val allBufferNrs = BufferNr.listOf(numberOfSnapshots)
+        val takenBufferNrs = savedSnapshots.keySet
+        val freeBufferNr = allBufferNrs.find(bufferNr => !takenBufferNrs.contains(bufferNr))
+        MonadThrow[F].fromOption(freeBufferNr, SnapshotStoreError("Could not find a free key")).flatMap { bufferNr =>
+          val wasApplied = statements.insertRecord(key, bufferNr, snapshot)
+          wasApplied.flatMap { wasApplied =>
+            // TODO: consider adding circuit breaker here
+            if (wasApplied) ().pure[F] else save(key, snapshot)
+          }
+        }
+      }
+
+      def replace(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        insertSnapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val oldestSnapshot = savedSnapshots.toList.minByOption { case (_, (seqNr, timestamp)) => (seqNr, timestamp) }
+
+        MonadThrow[F].fromOption(oldestSnapshot, SnapshotStoreError("Could not find an oldest snapshot")).flatMap {
+          oldestSnapshot =>
+            val (bufferNr, (deleteSnapshot, _)) = oldestSnapshot
+            val wasApplied = statements.updateRecord(key, bufferNr, insertSnapshot, deleteSnapshot)
+            wasApplied.flatMap { wasApplied =>
+              // TODO: consider adding circuit breaker here
+              if (wasApplied) ().pure[F] else save(key, insertSnapshot)
+            }
+        }
+      }
+
+      def update(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        insertSnapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val sortedSnapshots = savedSnapshots.toList.sortBy { case (_, (seqNr, timestamp)) => (seqNr, timestamp) }
```

Comment:
What's the point in sorting snapshots if later sorted list will be used only once:

```
sortedSnapshots.find ...
```
?

Reply:
The point is to update the oldest one if several snapshots are found.